res/controllers/DJ TechTools-MIDI Fighter Twister-scripts.js

```diff
@@ -54,6 +54,157 @@ var MidiFighterTwister;
+
+            this.volume = [];
+            this.mute = [];
+            this.effect = [];
+            this.effectEnabled = [];
+            this.effectVolume = [];
+            this.effectReset = [];
+            for (let i = 0; i < 4; i++) {
+                this.volume[i] = new components.Encoder({
+                    group: `[Channel${deckNums[0]}_Stem${i + 1}]`,
+                    midi: [0xB0, this.midiModifier(0x20 + (4 * i))],
+                    key: "volume",
+                });
+                // TODO: change these into mute or reset buttons based on a
+                // setting?
+                this.mute[i] = new components.Button({
+                    group: `[Channel${deckNums[0]}_Stem${i + 1}]`,
+                    midi: [0xB1, this.midiModifier(0x20 + (4 * i))],
+                    key: "mute",
+                    colorKey: "color",
+                    colorMapper: new ColorMapper({
+                        0x0040f0: 1,   // Blue
+                        0x3ec8e6: 32,  // Celeste
+                        0x008080: 40,  // Teal
+                        0x2fb340: 50,  // Green
+                        0x3fff58: 60,  // Lime
+                        0xe9c600: 66,  // Yellow
+                        0xf07800: 74,  // Orange
+                        0xfd4a4a: 85,  // Red (was being mapped to orange, perceptually closer to red)
+                        0xb90908: 85,  // Red
+                        0xff00ff: 100, // Fuscia
+                        0xa400c0: 108, // Purple
+                    }),
+                    type: components.Button.prototype.types.toggle,
+                    outValueScale: function(_value) {
+                        return this.colorMapper.getValueForNearestColor(engine.getParameter(this.group, this.colorKey));
+                    },
+                    shift: function() {
+                        this.inKey = "volume_set_default";
+                        this.outKey = "volume_set_default";
+                        this.type = components.Button.prototype.types.push;
+                        this.disconnect();
+                        this.connect();
+                        this.trigger();
+                    },
+                    unshift: function() {
+                        this.inKey = this.key;
+                        this.outKey = this.key;
+                        this.type = components.Button.prototype.types.toggle;
+                        this.disconnect();
+                        this.connect();
+                        this.trigger();
+                    },
+                    output: function(value, _group, _control) {
+                        if (value) {
+                            // Dim the LED
+                            midi.sendShortMsg(0xB2, this.midi[1], 35);
+                        } else {
+                            // Brighten the LED
+                            midi.sendShortMsg(0xB2, this.midi[1], 47);
+                        }
```

Comment:
does this mean the twister can control color and brightness independently by mapping color and brightness to two different midi channels (0xB1 for color, 0xB2 for brightness)?

Reply:
No, as far as I can see in their documentation there are just two colors per actual color: a bright variant and a dim variant, but there is no independent brightness control, sadly.

Reply:
I see. And I assume the encoders can also be pressed in to act as buttons?

Reply:
Correct, each encoder can be turned (and does the 2-s complement velocity thing, there is apparently also a mode where you can change them to absolute mode too but it's only part of their proprietary configuration software and I couldn't figure out how to do it), and they can also be pressed. There are also two sets of LEDs per encoder, one is an 11-segment ring LED that is associated with turning the encoder (and can be set independently but has no colors or brightness that can be set as far as I can tell) and a single RGB led that is associated with the button press control. This one has independent brightness and colors.